### DISCUSSION TITLE
Attribute Resolution

### DISCUSSION THREAD

**Discussion by jonathantneal** [withastro/roadmap#105] [Proposal]
In Astro, I would like boolean and string attributes to resolve consistently and in accordance with web standards. Currently, attribute expressions are handled inconsistently, hindering the usage of string attributes, boolean attributes, and asynchronous values.

### Proposed Rules

1. If a boolean is passed to an attribute, Astro will treat it like a boolean attribute.
2. If a string, number, or bigint is passed to an attribute, Astro will treat it like a string attribute.
3. If null or undefined is passed to an attribute, Astro will consider that value a boolean false.
4. If a promise is passed to to an attribute, Astro will pass the awaited result thru these proposed rules.
5. If another object or function is passed to an attribute, Astro will pass the string-ified result thru these proposed rules.

### Example: Boolean `true` includes the Boolean Attribute

| Astro Source                | HTML Result        |
| :-------------------------- | :----------------- |
| `<p hidden={true} />`       | `<p hidden>`       |
| `<p unknown={true} />`      | `<p unknown>`      |
| `<p mylib-hidden={true} />` | `<p mylib-hidden>` |
| `<p data-hidden={true} />`  | `<p data-hidden>`  |

### Example: Boolean `false` omits the Boolean Attribute

| Astro Source                 | HTML Result |
| :--------------------------- | :---------- |
| `<p hidden={false} />`       | `<p>`       |
| `<p unknown={false} />`      | `<p>`       |
| `<p mylib-hidden={false} />` | `<p>`       |
| `<p data-hidden={false} />`  | `<p>`       |

### Example: String `"true"` is a String Attribute

| Astro Source                | HTML Result               |
| :-------------------------- | :------------------------ |
| `<p hidden="true" />`       | `<p hidden="true">`       |
| `<p unknown="true" />`      | `<p unknown="true">`      |
| `<p mylib-hidden="true" />` | `<p mylib-hidden="true">` |
| `<p data-hidden="true" />`  | `<p data-hidden="true">`  |

### Example: Empty String `""` is a String Attribute

| Astro Source                | HTML Result               |
| :-------------------------- | :------------------------ |
| `<p hidden="" />`           | `<p hidden="">`           |
| `<p unknown="" />`          | `<p unknown="">`          |
| `<p mylib-hidden="" />`     | `<p mylib-hidden="">`     |
| `<p data-hidden="" />`      | `<p data-hidden="">`      |

### What about attributes that use the `"true"` value?

Attributes that use the `"true"` string value should use the `"true"` string value. This is clearly documented for attributes like `spellcheck`, `aria-hidden`, and `contenteditable`, the last of which supports both boolean and string values.

> The element's [`spellcheck`](https://html.spec.whatwg.org/multipage/interaction.html#attr-spellcheck) content attribute must be set to the literal string `"true"`, otherwise it must be set to the literal string `"false"`.
> — [HTML Living Standard](https://html.spec.whatwg.org/multipage/interaction.html#spelling-and-grammar-checking:attr-spellcheck-9)

> The suggested mappings for true/false values in HTML 5 use [Keyword and enumerated attributes](https://www.w3.org/TR/html5/infrastructure.html#keywords-and-enumerated-attributes) with allowed values of true and false, instead of using the HTML 5 boolean value type.
> — [ARIA 1.1](https://www.w3.org/TR/wai-aria-1.1/#h-note-60)

> The `contenteditable` content attribute is an [enumerated attribute](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#enumerated-attribute) whose keywords are the empty string, true, and false. The empty string and the **true keyword** map to the true state. The **false keyword** maps to the false state.
> — [HTML Living Standard](https://html.spec.whatwg.org/multipage/interaction.html#contenteditable:enumerated-attribute)

### Why does this matter?

Previously, React made special-cases for known combinations of elements and attributes. This special-casing later caused harm to the web by blocking framework-less Custom Elements and accessible ARIA properties. We should follow the web standard, and we should learn from React.

### How does it work today?

Currently, we do not align with either this proposal or React. Currently, a boolean value will create a boolean attribute when it matches a list of presently known boolean attributes, and also when the attribute is a data attribute; which does not align with this proposal or React. Currently, an empty string will create a boolean attribute, which aligns with this proposal but not React. Currently, a promise value will not resolve, which does not align with this proposal and aligns with React. Currently, a boolean false, null, or undefined value will omit the attribute; which does align with this proposal and also React.

**Comment by natemoo-re** [comment]
I'm not sure if this is an intentional omission, but we should also resolve how to handle "empty" attributes, which is another place we differ from HTML due to aligning with React/JSX behavior.

Empty attributes are a special construct in the compiler, so we do have the information to track them and handle them properly. However, we currently convert them to `true` when they are used on `<Component>` nodes but `""` when used on `<element>` nodes. This makes the usage of empty attributes like `<p hidden />` a lossy operation in most cases.

To correct this, we've added a list of the spec'd "boolean" attributes to our serialization logic that will correct `hidden=""` back to `hidden`. This is not ideal for obvious reasons, so finalizing our behavior here would be great.

**Reply by jonathantneal** [reply]
Thank you for looking at this, @natemoo-re. :)

This is not an intentional omission. An _empty_ attribute in the context of expressions is a string attribute. May I update the copy to reflect this? I can add an edit note at the bottom so that your comment remains relevant.

You mentioned that `""` is converted into `true` when used on `<Component>` nodes, but this is not what I currently experience. Would you mind confirming whether this demonstrates the preserved empty value (`""`) or if the example is somehow faulty?

[StackBlitz Example](https://stackblitz.com/edit/github-w26sgw?file=src/pages/index.astro)

**Reply by natemoo-re** [reply]
Ah, slight misunderstanding there. `<Component hidden="">` is not converted to `<Component hidden={true}>`. If you pass `<Component hidden="">` we respect it as literally authored. However, `<Component hidden>` is an _empty attribute_ which is interpreted (cast) differently for `<element>` versus `<Component>`.

[StackBlitz Example](https://stackblitz.com/edit/github-w26sgw-rhhfa4?file=src%2Fpages%2Findex.astro)

_Note: "Empty Attribute" is an Astro construct that is specified by neither HTML nor JSX. The behavior outlined above was Astro's attempt at bridging the gap between how HTML interprets attributes that are present with no explicit value as `""` and how JSX interprets props that are present with no explicit value as `true`_

**Reply by jonathantneal** [reply]
Thank you for explaining this. May I add a note about it? That current behavior also seems fully in alignment with this proposal.

**Comment by tony-sull** [comment]
> 4. If a promise is passed to to an attribute, Astro will pass the awaited result thru these proposed rules.
> 5. If another object or function is passed to an attribute, Astro will pass the string-ified result thru these proposed rules.

Automatic Promise resolution 🪄!!!

Would it make sense for promises and functions to work similarly, either both using the result as-is or both stringifying the results?

**Reply by natemoo-re** [reply]
Totally on board for automatic `Promise` resolution for attributes, especially when we are serializing `<element>` nodes. 

But do we want to automatically `await` Promise values in all cases? If I pass in `<Component result={fetch(...)}`, I would expect `Astro.props.result` to be `Promise<Response>` instead of `Response`.

**Reply by jonathantneal** [reply]
I would expect it to resolve, as it would in other expressions. However, if it did not resolve, then I would expect it would not resolve in other expressions.

I think a realistic use case for the `fetch` response would be to so something with the response status or headers. I could also see `fetch().then(response => response.text())`.

**Comment by natemoo-re** [comment]
> 5. If another object or function is passed to an attribute, Astro will pass the string-ified result thru these proposed rules.

When you say "string-ified" are you referring to the result of `new String(value)`? So `{}` would become `[object Object]`? That certainly fits with [the default HTML behavior](https://stackblitz.com/edit/js-klw5kp?file=index.js) but might be a minor inconvenience for users.

**Reply by jonathantneal** [reply]
In this instance, this proposal codifies the existing behavior, which is to effectively run it thru `String()`. Developers can then use `toString()` within an object or function to provide specialized behavior here. The same technique is used in React (_especially by CSS-in-JS libraries 😅_), but it works much better and more consistently in Astro 😄. And you are correct that, without any special `toString()` behavior, we’d often see `[object Object]`.

**Comment by tony-sull** [comment]
After sitting with this one a bit more and doing a few quick comparisons of the different component frameworks, I gotta say I really like the simplicity of this proposal.  I basically read this as Astro saying,

> If you want a boolean attribute, give me a `Boolean`.  If you want a string/enum attribute.



I compared a few common cases in Astro, React, Preact, and Svelte (chart below).  Each one has a different take on how to attempt to intelligently handle different attributes, but every framework acts in different ways.

There are good reasons for each decision made here, but it just highlights how tricky something as simple as attributes can be to handle with any kinds of intelligent/magic rules for handling values.  As I see this PR, we'd basically be acknowledging how context-specific attributes can be and leaving it entirely up to the developer to decide whether they want a particular attribute to be treated as a boolean or an enum


|  Component Source | Astro@next HTML | React HTML |  Svelte HTML | Preact HTML |
--------------------- | ------------------ | ------------ | -------------- | ------------- |
| `<p hidden="true">`  | `<p hidden="true">` | `<p hidden>`  | `<p hidden>`    | `<p hidden="true">` |
| `<p hidden="false">` | `<p hidden="false">` | `<p hidden>` | `<p hidden>`   | `<p hidden="false">` |
| `<p hidden="">`           | `<p hidden>`               | `<p>`             | `<p>`                | `<p hidden>`              |
| `<p data-hidden>`    | `<p data-hidden>`      | `<p data-hidden="true">` | `<p data-hidden>` | `<p data-hidden="true">` |
| `<p data-hidden={true} />` | `<p data-hidden>` | `<p data-hidden="true">` | `<p data-hidden="true">` | `<p data-hidden="true">` |
| `<p vue-hidden>` | `<p vue-hidden>` | `<p>` | `<p vue-hidden>` | `<p vue-hidden="true">` |
| `<p vue-hidden={true}>` | <p vue-hidden="true">` | `<p>` | `<p vue-hidden="true">` | `<p vue-hidden="true">` |


### Takeaway

That chart is a bit of a mess, sorry!  There's some really interesting differences though, and a few that I wouldn't have expected!  I think the direction of this PR goes a long way towards making it more understandable at least for Astro components - the framework will always respect the data type given for an attribute.

**Comment by TechAkayy** [comment]
**Attribute/prop Bindings**

For props/attributes, it might be a good idea to support expressions within quotes as well. This will cover all the above rules proposed for boolean. 

For eg, in [Svelte](https://svelte.dev/docs#template-syntax-attributes-and-props), both `<button disabled={!clickable}>...</button>` && `<button disabled="{!clickable}">...</button>` works the same. The second pattern is generally used when concating dynamic data with static, for eg 

> `<a href="page/{name}">page {name}</a>`

In [Vue](https://vuejs.org/guide/essentials/template-syntax.html#using-javascript-expressions), all attributes have to be strings. So, the above example will look like this 

> ```<a href="`page/${name}`">page {name}</a>```

In Astro, currently, assigning an expression to an attribute/prop is like this. 

> ```<a href={`page/${name}`}>page {name}</a>```

We can make this simpler in sync with Svelte's way. This way has a huge tooling advantage!

I'm not creating a separate rfc for this, thought it might make sense to cover in this one. Thank you!

**Reply by Tc-001** [reply]
The issue is that this breaks https://alpinejs.dev/, which is an alternative way to have reactive web websites without a framework, and it uses a copious amount of curly braces. (e.x. `x-data="{foo: 1, bar: 2}"` to define the initial state)

Because Astro is/was forked mainly from svelte, the functionality was there, but was removed to allow alpine to work without needing specific escaping (and maybe intelisense did not like it, not sure)

I could see something like `page=$"page/{name}"` working, but that would require a separate RFC

**Reply by TechAkayy** [reply]
Interesting to learn astro's alpine connection, but I'm not sure how the svelte way breaks alpine, will look at it with fresh eyes in the morning. If you can provide a simple example that would be very helpful. Thanks in advance! 

That said, I just learnt from our discord server that there is an undocumented astro feature where template literals can be used directly as expression in attribute binding without enclosing them with {}. So basically, my above example can be written in two forms and both works in astro it seems. 

Documented form (https://docs.astro.build/en/core-concepts/astro-components/#dynamic-attributes): 
> ```<a href={`page/${name}`}>page {name}</a>```

Undocumented form (vs-code extn update also pending)(https://discord.com/channels/830184174198718474/1009490715329437766/1009661762020327477)
>``` <a href=`page/${name}`>page {name}</a>```

The undocumented form feels a little hacky to my eyes, but I'm hoping it has been thought-through! Cheers!

**Reply by Tc-001** [reply]
Basically in svelte/your proposal `x-data="{foo:bar}"` would be interpreted as server-side and would be parsed (probably with errors) on build, when alpine expects the actual `{foo: bar}` string

But the `<a href=\`page/${name}`\>` seems to solve it quite nicely with ways familiar to a js dev, it definitely needs to be documented though

**Reply by TechAkayy** [reply]
Gotcha, thxs!

**Reply by lloydjatkinson** [reply]
Maybe this isn't the place for this but from what I can see having a single way of doing this, aka the shown Astro code, seems more uniform than providing two ways of binding? I can imagine the IDE tooling and compiler would need extra work to accommodate both possibly for unclear advantage? I think as it stands it's closer to JSX, which makes adopting Astro for other framework users that much simpler.

**Reply by jonathantneal** [reply]
Yes, this discussion was intended to give Astro consistent behavior. I’m not aware if that ended up happening or not.

Here is the key point from the original post:

> Currently, we do not align with either this proposal or React. Currently, a boolean value will create a boolean attribute when it matches a list of presently known boolean attributes, and also when the attribute is a data attribute; which does not align with this proposal or React. Currently, an empty string will create a boolean attribute, which aligns with this proposal but not React. Currently, a promise value will not resolve, which does not align with this proposal and aligns with React. Currently, a boolean false, null, or undefined value will omit the attribute; which does align with this proposal and also React.

**Comment by delucis** [comment]
Noting that this was resolved in Astro v5. https://docs.astro.build/en/guides/upgrade-to/v5/#changed-non-boolean-html-attribute-values
